File: strategy/generator/confirm/bb.py
from dataclasses import dataclass

from core.models.parameter import Parameter, StaticParameter
from core.models.smooth import Smooth
from core.models.source import SourceType

from .base import Confirm, ConfirmType


@dataclass(frozen=True)
class BbConfirm(Confirm):
    type: Confirm = ConfirmType.BbC
    source: Parameter = StaticParameter(SourceType.CLOSE)
    smooth: Parameter = StaticParameter(Smooth.SMA)
    period: Parameter = StaticParameter(20.0)
    factor: Parameter = StaticParameter(2.0)

File: ta_lib/strategies/confirm/src/bb.rs
use base::prelude::*;
use core::prelude::*;
use timeseries::prelude::*;
use volatility::bb;

pub struct BbConfirm {
    smooth: Smooth,
    period: usize,
    factor: f32,
}

impl BbConfirm {
    pub fn new(smooth: Smooth, period: f32, factor: f32) -> Self {
        Self {
            smooth,
            period: period as usize,
            factor,
        }
    }
}

impl Confirm for BbConfirm {
    fn lookback(&self) -> usize {
        self.period
    }

    fn filter(&self, data: &OHLCVSeries) -> (Series<bool>, Series<bool>) {
        let close = data.close();
        let prev_close = close.shift(1);
        let (upper_bb, _, lower_bb) = bb(close, self.smooth, self.period, self.factor);

        (
            close.sgt(&lower_bb) & prev_close.slt(&lower_bb),
            close.slt(&upper_bb) & prev_close.sgt(&upper_bb),
        )
    }
}
